chore(ci): correctly deploy code and docs on version branches and tags

gkalpak opened the pull request:
Previously, the generated build artifacts and docs were only deployed for builds associated with the master branch. There was also a `latest` branch mentioned in the config, but there is normally no such branch, so this had no effect.

This commit fixes the rules so that deployments happen when necessary.
More specifically:
- The `deploy-code` job now runs for builds associated with:
  - The master branch.
  - The stable branch (i.e. the branch from which the version tagged as `@latest` on npm is released).
  - Tags of the form `v1.X.Y(-Z)`. (This also required configuring CircleCI to run builds for git tags, which does not happen by default.)
- The `deploy-docs` job now runs for builds associated with:
  - The stable branch (i.e. the branch from which the version tagged as `@latest` on npm is released).

The new rules for when deployments should take place are based on the logic previously in [.travis.yml][1] and [scripts/travis/build.sh][2] (from before we switched from Travis to CircleCI).

[1]: https://github.com/angular/angular.js/blob/974700af7c17f07eb217d698dcfaf16796d1b263/.travis.yml#L54-L103
[2]: https://github.com/angular/angular.js/blob/974700af7c17f07eb217d698dcfaf16796d1b263/scripts/travis/build.sh#L66-L101

gkalpak committed:
chore(package.json): update docs app to use version 1.8 of AngularJS

As mentioned in `RELEASE.md`, now that the [CDN][1] has been updated
with the 1.8.0 version, it is safe to bump the value of the
`branchVersion` property in `package.json` to `^1.8.0`. This will cause
the docs app to use the latest version, namely 1.8.0.

[1]: https://ajax.googleapis.com/ajax/libs/angularjs/1.8.0/angular.js

gkalpak committed:
chore(ci): correctly deploy code and docs on version branches and tags

Previously, the generated build artifacts and docs were only deployed
for builds associated with the master branch. There was also a `latest`
branch mentioned in the config, but there is normally no such branch, so
this had no effect.

This commit fixes the rules so that deployments happen when necessary.
More specifically:
- The `deploy-code` job now runs for builds associated with:
  - The master branch.
  - The stable branch (i.e. the branch from which the version tagged as
    `@latest` on npm is released).
  - Tags of the form `v1.X.Y(-Z)`. (This also required configuring
    CircleCI to run builds for git tags, which does not happen by
    default.)
- The `deploy-docs` job now runs for builds associated with:
  - The stable branch (i.e. the branch from which the version tagged as
    `@latest` on npm is released).

The new rules for when deployments should take place are based on the
logic previously in [.travis.yml][1] and [scripts/travis/build.sh][2]
(from before we switched from Travis to CircleCI).

[1]: https://github.com/angular/angular.js/blob/974700af7c1/.travis.yml#L54-L103
[2]: https://github.com/angular/angular.js/blob/974700af7c1/scripts/travis/build.sh#L66-L101

Splaktar reviewed (approved):
LGTM

g-patel commented:
@gkalpak Thanks much for doing this work for the old framework that is Angular1.x. 
Are we waiting on some before it can be merged? I don't want to be pushy, but I am curious as to the [1.8 version of the docs](https://code.angularjs.org/1.8.0/docs/misc/version-support-status) is what is referenced from the main site, and I believe is treated as source of truth for LTS/support.

Splaktar commented:
@g-patel most like we'll need @petebacondarwin to review this and approve it as well.

FYI: the [AngularJS LTS blog post](https://blog.angular.io/stable-angularjs-and-long-term-support-7e077635ee9c) was updated to include the Covid-19 extension of the LTS.

g-patel commented:
> @g-patel most like we'll need @petebacondarwin to review this and approve it as well.
> 
> FYI: the [AngularJS LTS blog post](https://blog.angular.io/stable-angularjs-and-long-term-support-7e077635ee9c) was updated to include the Covid-19 extension of the LTS.

Thanks for reply @Splaktar. Good to know that the blog post has been updated. 👍

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -292,44 +369,70 @@ jobs:
+      - custom_attach_workspace
+      - init_environment
+      - skip_unless_stable_branch
+      - run: yarn grunt prepareDeploy
```
Is this needed since the `deploy-docs` job requires the `prepare-deployment` job?

gkalpak replied:
Yes, it is still necessary, because we are not persisting all affected files to the workspace.

Currently, the `prepareDeploy` command will affect files in the `deploy/` and `scripts/docs.angularjs.org-firebase/` directories and also create a `firebase.json` file at the root directory, but we [only persist the `deploy/` directory][1].

I guess we could persist the whole `ng/` directory and not require this, though. I'll update the PR.

[1]: https://github.com/angular/angular.js/blob/295213df9537666254626ffe3e4a6013122e4802/.circleci/config.yml#L265

petebacondarwin reviewed (approved):
one minor question - otherwise LGTM

gkalpak committed:
chore(ci): avoid unnecessarily running `grunt prepareDeploy` in `deploy-docs` CI job

Previously, the `grunt prepareDeploy` command was run in both the
`prepare-deployment` and `deploy-docs` CI jobs. The reason was that not
all files affected by `grunt prepareDeploy` were persisted to the
workspace across jobs.

More specifically, the command would affect files in the `deploy/` and
`scripts/docs.angularjs.org-firebase/` directories and also create a
`firebase.json` file at the root directory, but only the `deploy/`
directory was [persisted to the workspace][1].

This commit avoids unnecessarily running the `grunt prepareDeploy`
command in the `deploy-docs` CI job by ensuring that all affected files
will be persisted to the workspace in the `prepare-deployment` CI job,
which always runs before `deploy-docs`.

[1]: https://github.com/angular/angular.js/blob/295213df953766625462/.circleci/config.yml#L265
